trainer/trainer.py

```diff
@@ -0,0 +1,322 @@
+            print(stage)
+            self._init_stage_(self.stages[stage])
+            # The first time we get here, if we are resuming training, we need to restore the dataset state
+            # Restore state and then don't do it again:
+            if self.restore_datasets:
+                for dataset in self.stages[stage].datasets.keys():
+                    mystate: DatasetState = self.state_tracker.get_dataset(dataset)
+                    self.dataset_objects[dataset].restore_state(mystate)
+                    print("Restoring dataset:", dataset, "to seed:", mystate.seed, "line:", mystate.line, "epoch:", mystate.epoch)
+                self.restore_datasets = False
+            self.state_tracker.update_stage(stage)
+            self.train_stage(self.stages[stage])
+
+
+    def _init_stage_(self, stage: Stage): #@TODO make the stupid stage a full object so i can have proper attributes
+        '''Init a certain stage of the training'''
+        for dataset in stage.datasets.keys():
+            self.dataset_objects[dataset].set_weight(stage.datasets[dataset])
+            self.dataset_objects[dataset].set_max_epoch(inf)
+            self.dataset_objects[dataset].reset_epoch()
+        self.dataset_objects[stage.until_dataset].set_max_epoch(stage.until_epoch)
+
+    def _restore_(self) -> None:
+        """Restores the training state from a state tracker"""
+        # Skip training stages based on if we are resuming the training
+        if self.state_tracker.stage is not None:
+            resume_stage_idx = self.stage_names.index(self.state_tracker.stage)
+            print("Skipping training stages", self.stage_names[0:resume_stage_idx], "as we are resuming training")
+            self.stage_names = self.stage_names[resume_stage_idx:]
+            # Restore random state
+            self.state_tracker.restore_random_state()
+
+    def train_stage(self, stage):
+        '''Trains up to a training stage'''
+        # Make sure that we check the stopping criteria before starting to train so we don't continue
+        # training after we have finished everything else first.
+        stop_training: bool = self.dataset_objects[stage.until_dataset].epoch >= stage.until_epoch
+        while not stop_training:
+            batch = []
+            for dataset in stage.datasets:
+                _, lines = self.dataset_objects[dataset].get()
+                #print(epoch, dataset, stage.until_dataset, stage.until_epoch, len(lines))
+                batch.extend(lines)
+            # Shuffle the batch
+            random.shuffle(batch)
+            # Uppercase randomly
+            batch = [x.upper() if random.random() < self.uppercase_ratio else x for x in batch]
+            self.trainer.stdin.writelines(batch)
+            self.state_tracker.update_seed()
+            # Termination condition, check if we finished training.
+            stop_training = self.dataset_objects[stage.until_dataset].epoch >= stage.until_epoch
```

Comment:
Maybe we need a `self.trainer.wait()` somewhere here? The `communicate` function takes complete control of the suprocess but it seems to not fit the purpose of this function.

Reply:
I think the wait should avoid the trainer.py to end before marian won't kill marian if CTRl+C. Maybe try to catch the CTRL+C exception and then wait()?

Reply:
I think it would be:
```python
try:
     # train loop
except KeyboardInterrupt:
     trainer.wait()
     sys.exit()
```

Reply:
Add a `trainer.stdin.close()` before that `trainer.wait()` and that should stop the trainer cleanly. Alternatively you can use `trainer.kill()` to terminate it immediately. 

What I often do is:
```py
try:
  # train loop
finally:
  trainer.stdin.close()
  trainer.wait()
```

That way, in both the interrupted and the normal case you end with stopping the trainer cleanly. Although in this use case I'm not sure whether `trainer.kill()` might be helpful in an except so you don't have to wait too long.

Reply:
Should be fixed